contracts/SwapAndCall.sol

```diff
@@ -0,0 +1,80 @@
+pragma solidity ^0.5.0;
+
+import '@openzeppelin/contracts/token/ERC20/IERC20.sol';
+import '@openzeppelin/contracts/token/ERC20/SafeERC20.sol';
+import 'hardlydifficult-ethereum-contracts/contracts/proxies/CallContract.sol';
+
+
+/**
+ * @notice call any contract(s) and then refund the remaining balance (if any)
+ * Idea from https://etherscan.io/address/0x11111254369792b2Ca5d084aB5eEA397cA8fa48B#code
+ */
+contract SwapAndCall
+{
+  using CallContract for address;
+  using SafeERC20 for IERC20;
+
+  // Accept ETH from other contracts
+  function() external payable {}
+
+  /**
+   * @notice Make a series of calls and then refunds any funds remaining.
+   * @param _contractsAndTokensToRefund The contract address for each call to make.
+   * Followed by the address of any ERC-20 tokens to refund at the end.
+   * ETH is always refunded.  If no additional token is required this may be empty.
+   * @param _callDataConcat Call data for each call, appended into a single bytes field.
+   * @param _startPositionsAndValues The start position of callData for each call after
+   * the first (assumed to be 0).
+   * Followed by the call value to include with each call.
+   * @dev Variables are reused to avoid a `stack too deep` error.
+   */
+  function swapAndCall(
+    address[] calldata _contractsAndTokensToRefund,
+    bytes calldata _callDataConcat,
+    uint[] calldata _startPositionsAndValues
+  ) external payable
+  {
+    uint callCountMinusOne = _startPositionsAndValues.length / 2;
+    for (uint i = 0; i <= callCountMinusOne; i++)
+    {
+      uint startPosition = 0;
+      uint length;
+      if(i > 0)
+      {
+        startPosition = _startPositionsAndValues[i - 1];
+      }
+      if(i == callCountMinusOne)
+      {
+        length = _callDataConcat.length;
+      }
+      else
+      {
+        length = _startPositionsAndValues[i];
+      }
+      length -= startPosition;
+      _contractsAndTokensToRefund[i]._callByPosition(
+        _callDataConcat,
+        startPosition,
+        length,
+        _startPositionsAndValues[callCountMinusOne + i]
+      );
+    }
+
+    uint amount = address(this).balance;
+    if(amount > 0)
+    {
+      msg.sender.transfer(amount);
+    }
+
+    uint max = _contractsAndTokensToRefund.length;
+    for (uint i = callCountMinusOne; i < max; i++)
```

Comment:
So with the example I gave above in which there are a total of 4 calls to make, `i` here would start at `4`, which would be the first of the addresses for tokens to refund. (indexes 0-3 would be the contracts to call)
Sound right?